src/optimize/watch/watch_cache.ts

```diff
@@ -0,0 +1,173 @@
+
+interface Params {
+  log: (tags: string[], data: string) => void;
+  outputPath: string;
+  dllsPath: string;
+  cachePath: string;
+}
+
+interface WatchCacheStateContent {
+  optimizerConfigSha?: string;
+  yarnLockSha?: string;
+}
+
+export class WatchCache {
+  private readonly log: Params['log'];
+  private readonly outputPath: Params['outputPath'];
+  private readonly dllsPath: Params['dllsPath'];
+  private readonly cachePath: Params['cachePath'];
+  private readonly cacheState: WatchCacheStateContent;
+  private statePath: string;
+  private diskCacheState: WatchCacheStateContent;
+  private isInitialized: boolean;
+
+  constructor(params: Params) {
+    this.log = params.log;
+    this.outputPath = params.outputPath;
+    this.dllsPath = params.dllsPath;
+    this.cachePath = params.cachePath;
+
+    this.isInitialized = false;
+    this.statePath = '';
+    this.cacheState = {};
+    this.diskCacheState = {};
+    this.cacheState.yarnLockSha = '';
+    this.cacheState.optimizerConfigSha = '';
+  }
+
+  public async tryInit() {
+    if (!this.isInitialized) {
+      this.statePath = resolve(this.outputPath, 'watch_optimizer_cache_state.json');
+      this.diskCacheState = await this.read();
+      this.cacheState.yarnLockSha = await this.buildYarnLockSha();
+      this.cacheState.optimizerConfigSha = await this.buildOptimizerConfigSha();
+      this.isInitialized = true;
+    }
+  }
+
+  public async tryReset() {
+    await this.tryInit();
+
+    if (!this.isResetNeeded()) {
+      return;
+    }
+
+    await this.reset();
+  }
+
+  public async reset() {
+    this.log(['info', 'optimize:watch_cache'], 'The optimizer watch cache will reset');
+
+    // start by deleting the state file to lower the
+    // amount of time that another process might be able to
+    // successfully read it once we decide to delete it
+    await del(this.statePath);
+
+    // first delete some empty folder that left
+    // from any previous cache reset action
+    await deleteEmpty(`${this.cachePath}`);
+
+    // delete everything in optimize/.cache directory
+    // except ts-node
+    await del(await globby([`${this.cachePath}`, `!${this.cachePath}/ts-node/**`], { dot: true }));
+
+    // delete dlls
+    await del(this.dllsPath);
+    await mkdirpAsync(this.dllsPath);
+
+    // re-write new cache state file
+    await this.write();
+
+    this.log(['info', 'optimize:watch_cache'], 'The optimizer watch cache has reset');
+  }
+
+  private async buildShaWithMultipleFiles(filePaths: string[]) {
+    const shaHash = createHash('sha1');
+
+    for (const filePath of filePaths) {
+      try {
+        shaHash.update(await readAsync(filePath), 'utf8');
+      } catch (e) {
+        /* no-op */
+      }
+    }
+
+    return shaHash.digest('hex');
+  }
+
+  private async buildYarnLockSha() {
+    const kibanaYarnLock = resolve(__dirname, '../../../yarn.lock');
+    const xpackYarnLock = resolve(__dirname, '../../../x-pack/yarn.lock');
```

Comment:
i think we can ignore this one.  in dev the top level yarn.lock would cover both, and in prod the path would be wrong?

Reply:
The watch catch doesn't run in prod @jbudz 😃 I think we need to also keep track of the x-pack yarn lock to rebuild the dll if some of the dependencies used by x-pack just changed the version specifically for the x-pack package right?

Reply:
Ok right now with the last changes to the bootstrap changes I can drop the usage of the x-pack/yarn.lock as we only have one yarn.lock for the entire project